// igniteui-angular/esm5/lib/drop-down/drop-down-item.base.js
import { __decorate, __param } from "tslib";
import { IGX_DROPDOWN_BASE } from './drop-down.common';
import { Directive, Input, HostBinding, HostListener, ElementRef, Optional, Inject, DoCheck, Output, EventEmitter } from '@angular/core';
import { IgxSelectionAPIService } from '../core/selection';
import { showMessage } from '../core/deprecateDecorators';
import { IgxDropDownGroupComponent } from './drop-down-group.component';
var NEXT_ID = 0;
var warningShown = false;
/**
 * An abstract class defining a drop-down item:
 * With properties / styles for selection, highlight, height
 * Bindable property for passing data (`value: any`)
 * Parent component (has to be used under a parent with type `IDropDownBase`)
 * Method for handling click on Host()
 */
var IgxDropDownItemBaseDirective = /** @class */ (function () {
    function IgxDropDownItemBaseDirective(dropDown, elementRef, group, selection) {
        this.dropDown = dropDown;
        this.elementRef = elementRef;
        this.group = group;
        this.selection = selection;
        /**
         * @hidden
         */
        this._focused = false;
        this._selected = false;
        this._index = null;
        this._disabled = false;
        /**
         * Sets/gets the `id` of the item.
         * ```html
         * <igx-drop-down-item [id] = 'igx-drop-down-item-0'></igx-drop-down-item>
         * ```
         * ```typescript
         * let itemId =  this.item.id;
         * ```
         * @memberof IgxSelectItemComponent
         */
        this.id = "igx-drop-down-item-" + NEXT_ID++;
        /**
         * @hidden
         */
        this.selectedChange = new EventEmitter();
        /**
         * Gets/sets the `role` attribute of the item. Default is 'option'.
         *
         * ```html
         *  <igx-drop-down-item [role]="customRole"></igx-drop-down-item>
         * ```
         */
        this.role = 'option';
    }
    Object.defineProperty(IgxDropDownItemBaseDirective.prototype, "hasIndex", {
        get: function () {
            return this._index !== null && this._index !== undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxDropDownItemBaseDirective.prototype, "itemID", {
        /**
         * @hidden @internal
         */
        get: function () {
            return this;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxDropDownItemBaseDirective.prototype, "index", {
        /**
         * The data index of the dropdown item.
         *
         * ```typescript
         * // get the data index of the selected dropdown item
         * let selectedItemIndex = this.dropdown.selectedItem.index
         * ```
         */
        get: function () {
            if (this._index === null) {
                warningShown = showMessage('IgxDropDownItemBaseDirective: Automatic index is deprecated.' +
                    'Bind in the template instead using `<igx-drop-down-item [index]="i"` instead.`', warningShown);
                return this.itemIndex;
            }
            return this._index;
        },
        set: function (value) {
            this._index = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxDropDownItemBaseDirective.prototype, "itemStyle", {
        /**
         * @hidden @internal
         */
        get: function () {
            return !this.isHeader;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxDropDownItemBaseDirective.prototype, "itemStyleCosy", {
        /**
         * @hidden @internal
         */
        get: function () {
            return this.dropDown.displayDensity === 'cosy' && !this.isHeader;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxDropDownItemBaseDirective.prototype, "itemStyleCompact", {
        /**
         * @hidden @internal
         */
        get: function () {
            return this.dropDown.displayDensity === 'compact' && !this.isHeader;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxDropDownItemBaseDirective.prototype, "selected", {
        /**
         * Sets/Gets if the item is the currently selected one in the dropdown
         *
         * ```typescript
         *  let mySelectedItem = this.dropdown.selectedItem;
         *  let isMyItemSelected = mySelectedItem.selected; // true
         * ```
         *
         * Two-way data binding
         * ```html
         * <igx-drop-down-item [(selected)]='model.isSelected'></igx-drop-down-item>
         * ```
         */
        get: function () {
            return this._selected;
        },
        set: function (value) {
            if (this.isHeader) {
                return;
            }
            this._selected = value;
            this.selectedChange.emit(this._selected);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxDropDownItemBaseDirective.prototype, "focused", {
        /**
         * Sets/gets if the given item is focused
         * ```typescript
         *  let mySelectedItem = this.dropdown.selectedItem;
         *  let isMyItemFocused = mySelectedItem.focused;
         * ```
         */
        get: function () {
            return this.isSelectable && this._focused;
        },
        /**
         * ```html
         *  <igx-drop-down-item *ngFor="let item of items" focused={{!item.focused}}>
         *      <div>
         *          {{item.field}}
         *      </div>
         *  </igx-drop-down-item>
         * ```
         */
        set: function (value) {
            this._focused = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxDropDownItemBaseDirective.prototype, "headerClassCosy", {
        /**
         * @hidden @internal
         */
        get: function () {
            return this.isHeader && this.dropDown.displayDensity === 'cosy';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxDropDownItemBaseDirective.prototype, "headerClassCompact", {
        /**
         * @hidden @internal
         */
        get: function () {
            return this.isHeader && this.dropDown.displayDensity === 'compact';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxDropDownItemBaseDirective.prototype, "disabled", {
        /**
         * Sets/gets if the given item is disabled
         *
         * ```typescript
         *  // get
         *  let mySelectedItem = this.dropdown.selectedItem;
         *  let myItemIsDisabled = mySelectedItem.disabled;
         * ```
         *
         * ```html
         *  <igx-drop-down-item *ngFor="let item of items" disabled={{!item.disabled}}>
         *      <div>
         *          {{item.field}}
         *      </div>
         *  </igx-drop-down-item>
         * ```
         * **NOTE:** Drop-down items inside of a disabled `IgxDropDownGroup` will always count as disabled
         */
        get: function () {
            return this.group ? this.group.disabled || this._disabled : this._disabled;
        },
        set: function (value) {
            this._disabled = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxDropDownItemBaseDirective.prototype, "itemIndex", {
        /**
         * Gets item index
         * @hidden @internal
         */
        get: function () {
            return this.dropDown.items.indexOf(this);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxDropDownItemBaseDirective.prototype, "elementHeight", {
        /**
         * Gets item element height
         * @hidden @internal
         */
        get: function () {
            return this.elementRef.nativeElement.clientHeight;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxDropDownItemBaseDirective.prototype, "element", {
        /**
         * Get item html element
         * @hidden @internal
         */
        get: function () {
            return this.elementRef;
        },
        enumerable: true,
        configurable: true
    });
    IgxDropDownItemBaseDirective.prototype.ngDoCheck = function () {
        if (this._selected) {
            var dropDownSelectedItem = this.dropDown.selectedItem;
            if (!dropDownSelectedItem) {
                this.dropDown.selectItem(this);
            }
            else if (this.hasIndex
                ? this._index !== dropDownSelectedItem.index || this.value !== dropDownSelectedItem.value :
                this !== dropDownSelectedItem) {
                this.dropDown.selectItem(this);
            }
        }
    };
    Object.defineProperty(IgxDropDownItemBaseDirective.prototype, "isSelectable", {
        /** Returns true if the items is not a header or disabled  */
        get: function () {
            return !(this.disabled || this.isHeader);
        },
        enumerable: true,
        configurable: true
    });
    /** If `allowItemsFocus` is enabled, keep the browser focus on the active item */
    IgxDropDownItemBaseDirective.prototype.ensureItemFocus = function () {
        if (this.dropDown.allowItemsFocus) {
            var focusedItem = this.dropDown.items.find(function (item) { return item.focused; });
            if (!focusedItem) {
                return;
            }
            focusedItem.element.nativeElement.focus({ preventScroll: true });
        }
    };
    /**
     * @hidden
     * @internal
     */
    IgxDropDownItemBaseDirective.prototype.clicked = function (event) {
    };
    IgxDropDownItemBaseDirective.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [IGX_DROPDOWN_BASE,] }] },
        { type: ElementRef },
        { type: IgxDropDownGroupComponent, decorators: [{ type: Optional }] },
        { type: IgxSelectionAPIService, decorators: [{ type: Optional }, { type: Inject, args: [IgxSelectionAPIService,] }] }
    ]; };
    __decorate([
        HostBinding('attr.id'),
        Input()
    ], IgxDropDownItemBaseDirective.prototype, "id", void 0);
    __decorate([
        Input()
    ], IgxDropDownItemBaseDirective.prototype, "index", null);
    __decorate([
        Input()
    ], IgxDropDownItemBaseDirective.prototype, "value", void 0);
    __decorate([
        HostBinding('class.igx-drop-down__item')
    ], IgxDropDownItemBaseDirective.prototype, "itemStyle", null);
    __decorate([
        HostBinding('class.igx-drop-down__item--cosy')
    ], IgxDropDownItemBaseDirective.prototype, "itemStyleCosy", null);
    __decorate([
        HostBinding('class.igx-drop-down__item--compact')
    ], IgxDropDownItemBaseDirective.prototype, "itemStyleCompact", null);
    __decorate([
        Input(),
        HostBinding('attr.aria-selected'),
        HostBinding('class.igx-drop-down__item--selected')
    ], IgxDropDownItemBaseDirective.prototype, "selected", null);
    __decorate([
        Output()
    ], IgxDropDownItemBaseDirective.prototype, "selectedChange", void 0);
    __decorate([
        HostBinding('class.igx-drop-down__item--focused')
    ], IgxDropDownItemBaseDirective.prototype, "focused", null);
    __decorate([
        Input(),
        HostBinding('class.igx-drop-down__header')
    ], IgxDropDownItemBaseDirective.prototype, "isHeader", void 0);
    __decorate([
        HostBinding('class.igx-drop-down__header--cosy')
    ], IgxDropDownItemBaseDirective.prototype, "headerClassCosy", null);
    __decorate([
        HostBinding('class.igx-drop-down__header--compact')
    ], IgxDropDownItemBaseDirective.prototype, "headerClassCompact", null);
    __decorate([
        Input(),
        HostBinding('attr.aria-disabled'),
        HostBinding('class.igx-drop-down__item--disabled')
    ], IgxDropDownItemBaseDirective.prototype, "disabled", null);
    __decorate([
        Input(),
        HostBinding('attr.role')
    ], IgxDropDownItemBaseDirective.prototype, "role", void 0);
    __decorate([
        HostListener('click', ['$event'])
    ], IgxDropDownItemBaseDirective.prototype, "clicked", null);
    IgxDropDownItemBaseDirective = __decorate([
        Directive({
            selector: '[igxDropDownItemBase]'
        }),
        __param(0, Inject(IGX_DROPDOWN_BASE)),
        __param(2, Optional()),
        __param(3, Optional()), __param(3, Inject(IgxSelectionAPIService))
    ], IgxDropDownItemBaseDirective);
    return IgxDropDownItemBaseDirective;
}());
export { IgxDropDownItemBaseDirective };
