// igniteui-angular/esm5/lib/grids/moving/moving.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * This enumeration is used to configure whether the pinning possition is set before or after
 * the target. DropPosition.None is acting like DropPosition.AfterDropTarget.
 */
export var DropPosition;
(function (DropPosition) {
    DropPosition[DropPosition["BeforeDropTarget"] = 0] = "BeforeDropTarget";
    DropPosition[DropPosition["AfterDropTarget"] = 1] = "AfterDropTarget";
    DropPosition[DropPosition["None"] = 2] = "None";
})(DropPosition || (DropPosition = {}));
/**
 * @hidden
 * @internal
 */
var IgxColumnMovingService = /** @class */ (function () {
    function IgxColumnMovingService() {
    }
    Object.defineProperty(IgxColumnMovingService.prototype, "column", {
        get: function () {
            return this._column;
        },
        set: function (val) {
            if (val) {
                this._column = val;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgxColumnMovingService.prototype, "icon", {
        get: function () {
            return this._icon;
        },
        set: function (val) {
            if (val) {
                this._icon = val;
            }
        },
        enumerable: true,
        configurable: true
    });
    IgxColumnMovingService.ɵprov = i0.ɵɵdefineInjectable({ factory: function IgxColumnMovingService_Factory() { return new IgxColumnMovingService(); }, token: IgxColumnMovingService, providedIn: "root" });
    IgxColumnMovingService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], IgxColumnMovingService);
    return IgxColumnMovingService;
}());
export { IgxColumnMovingService };
