// igniteui-angular/esm2015/lib/directives/for-of/base.helper.component.js
import { __decorate } from "tslib";
import { EventEmitter, Output, HostListener, ElementRef, ChangeDetectorRef, OnDestroy, Directive, AfterViewInit } from '@angular/core';
let VirtualHelperBaseDirective = class VirtualHelperBaseDirective {
    constructor(elementRef, cdr) {
        this.elementRef = elementRef;
        this.cdr = cdr;
        this.scrollAmount = 0;
        this._size = 0;
        this._afterViewInit = false;
    }
    ngAfterViewInit() {
        this._afterViewInit = true;
    }
    onScroll(event) {
        this.scrollAmount = event.target.scrollTop || event.target.scrollLeft;
    }
    get nativeElement() {
        return this.elementRef.nativeElement;
    }
    ngOnDestroy() {
        this.destroyed = true;
    }
    set size(value) {
        if (this.destroyed) {
            return;
        }
        this._size = value;
        if (this._afterViewInit) {
            this.cdr.detectChanges();
        }
    }
    get size() {
        return this._size;
    }
    get scrollNativeSize() {
        const div = document.createElement('div');
        const style = div.style;
        style.width = '100px';
        style.height = '100px';
        style.position = 'absolute';
        style.top = '-10000px';
        style.top = '-10000px';
        style.overflow = 'scroll';
        document.body.appendChild(div);
        const scrollWidth = div.offsetWidth - div.clientWidth;
        document.body.removeChild(div);
        return scrollWidth ? scrollWidth + 1 : 1;
    }
};
VirtualHelperBaseDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: ChangeDetectorRef }
];
__decorate([
    HostListener('scroll', ['$event'])
], VirtualHelperBaseDirective.prototype, "onScroll", null);
VirtualHelperBaseDirective = __decorate([
    Directive({
        selector: '[igxVirtualHelperBase]'
    })
], VirtualHelperBaseDirective);
export { VirtualHelperBaseDirective };
