// igniteui-angular/esm5/lib/grids/common/enums.js
export var FilterMode;
(function (FilterMode) {
    FilterMode["quickFilter"] = "quickFilter";
    FilterMode["excelStyleFilter"] = "excelStyleFilter";
})(FilterMode || (FilterMode = {}));
export var GridSummaryPosition;
(function (GridSummaryPosition) {
    GridSummaryPosition["top"] = "top";
    GridSummaryPosition["bottom"] = "bottom";
})(GridSummaryPosition || (GridSummaryPosition = {}));
export var GridSummaryCalculationMode;
(function (GridSummaryCalculationMode) {
    GridSummaryCalculationMode["rootLevelOnly"] = "rootLevelOnly";
    GridSummaryCalculationMode["childLevelsOnly"] = "childLevelsOnly";
    GridSummaryCalculationMode["rootAndChildLevels"] = "rootAndChildLevels";
})(GridSummaryCalculationMode || (GridSummaryCalculationMode = {}));
export var GridKeydownTargetType;
(function (GridKeydownTargetType) {
    GridKeydownTargetType["dataCell"] = "dataCell";
    GridKeydownTargetType["summaryCell"] = "summaryCell";
    GridKeydownTargetType["groupRow"] = "groupRow";
    GridKeydownTargetType["hierarchicalRow"] = "hierarchicalRow";
})(GridKeydownTargetType || (GridKeydownTargetType = {}));
export var GridSelectionMode;
(function (GridSelectionMode) {
    GridSelectionMode["none"] = "none";
    GridSelectionMode["single"] = "single";
    GridSelectionMode["multiple"] = "multiple";
})(GridSelectionMode || (GridSelectionMode = {}));
export var ColumnDisplayOrder;
(function (ColumnDisplayOrder) {
    ColumnDisplayOrder["Alphabetical"] = "Alphabetical";
    ColumnDisplayOrder["DisplayOrder"] = "DisplayOrder";
})(ColumnDisplayOrder || (ColumnDisplayOrder = {}));
export var ColumnPinningPosition;
(function (ColumnPinningPosition) {
    ColumnPinningPosition[ColumnPinningPosition["Start"] = 0] = "Start";
    ColumnPinningPosition[ColumnPinningPosition["End"] = 1] = "End";
})(ColumnPinningPosition || (ColumnPinningPosition = {}));
export var RowPinningPosition;
(function (RowPinningPosition) {
    RowPinningPosition[RowPinningPosition["Top"] = 0] = "Top";
    RowPinningPosition[RowPinningPosition["Bottom"] = 1] = "Bottom";
})(RowPinningPosition || (RowPinningPosition = {}));
export var GridPagingMode;
(function (GridPagingMode) {
    GridPagingMode[GridPagingMode["local"] = 0] = "local";
    GridPagingMode[GridPagingMode["remote"] = 1] = "remote";
})(GridPagingMode || (GridPagingMode = {}));
